src/librustc/ty/sty.rs

```diff
@@ -1472,6 +1472,43 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
         }
     }
 
+    pub fn conservative_is_uninhabited(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
+        // Checks whether a type is definitely uninhabited. This is
+        // conservative: for some types that are uninhabited we return `false`,
+        // but we only return `true` for types that are definitely uninhabited.
+        match self.sty {
+            ty::Never => true,
+            ty::Adt(def, _) => {
+                // Any ADT is uninhabited if:
```

Comment:
Should probably add "either" to this, to indicate `a | b` (right now it could be read at a glance as `a & b`).